boot/zephyr/decompression.c

```diff
@@ -0,0 +1,987 @@
+                last_packet = true;
+            }
+
+            rc = compression->decompress(NULL, &buf[tmp_off], chunk_size, last_packet, &offset,
+                                         &output, &output_size);
+
+            if (rc) {
+                rc = BOOT_EBADSTATUS;
+                goto finish;
+            }
+
+            /* Copy data to secondary buffer for writing out */
+            while (output_size > 0) {
+                uint32_t data_size = (sizeof(decomp_buf) - decomp_buf_size);
+
+                if (data_size > output_size) {
+                    data_size = output_size;
+                }
+
+                memcpy(&decomp_buf[decomp_buf_size], output, data_size);
+                memmove(output, &output[data_size], output_size - data_size);
+
+                decomp_buf_size += data_size;
+                output_size -= data_size;
+
+                /* Write data out from secondary buffer when it is full */
+                if (decomp_buf_size == sizeof(decomp_buf)) {
+                    rc = flash_area_write(fap_dst, (off_dst + hdr->ih_hdr_size + write_pos),
+                                          decomp_buf, sizeof(decomp_buf));
+
+                    if (rc != 0) {
+                        rc = BOOT_EFLASH;
+                        goto finish;
+                    }
+
+                    write_pos += sizeof(decomp_buf);
+                    decomp_buf_size = 0;
+                }
+            }
+
+            tmp_off += chunk_size;
+        }
+
+        pos += copy_size;
+    }
+
+    /* Clean up decompression system */
+    (void)compression->deinit(NULL);
+
+    if (protected_tlv_size > 0) {
+        rc = boot_copy_protected_tlvs(hdr, fap_src, fap_dst, (off_dst + hdr->ih_hdr_size +
+                                                              write_pos), protected_tlv_size,
+                                      decomp_buf, sizeof(decomp_buf_size), &decomp_buf_size,
+                                      &tlv_write_size);
+
+        if (rc) {
+            goto finish;
+        }
+
+        write_pos += tlv_write_size;
+    }
+
+    tlv_write_size = 0;
+    rc = boot_copy_unprotected_tlvs(hdr, fap_src, fap_dst, (off_dst + hdr->ih_hdr_size +
+                                                            write_pos), unprotected_tlv_size,
+                                    decomp_buf, sizeof(decomp_buf_size), &decomp_buf_size,
+                                    &tlv_write_size);
+
+    if (rc) {
+        goto finish;
+    }
+
+    write_pos += tlv_write_size;
+
+    /* Check if we have unwritten data buffered up and, if so, write it out */
+    if (decomp_buf_size > 0) {
+        uint32_t write_padding_size = decomp_buf_size % write_alignment;
+
+        /* Check if additional write padding should be applied to meet the minimum write size */
+        if (write_padding_size) {
+            memset(&decomp_buf[decomp_buf_size], 0xff, write_padding_size);
```

Comment:
I know we are doing this for the flash devices at this point, but this should be erase_value rather than 0xff. Shouldn't it?

Reply:
Fixed